client/lib/i18n-utils/utils.js

```diff
@@ -110,18 +110,95 @@ export function getLocaleFromPath( path ) {
  * @returns {string} original path with new locale slug
  */
 export function addLocaleToPath( path, locale ) {
-	const urlParts = parse( path );
+	const urlParts = url.parse( path );
 	const queryString = urlParts.search || '';
 
 	return removeLocaleFromPath( urlParts.pathname ) + `/${ locale }` + queryString;
 }
 
-export function addLocaleToWpcomUrl( url, locale ) {
-	if ( locale && locale !== 'en' ) {
-		return url.replace( 'https://wordpress.com', 'https://' + locale + '.wordpress.com' );
+const localesWithBlog = [ 'en', 'ja', 'es', 'pt', 'fr', 'pt-br' ];
+const localesWithPrivacyPolicy = [ 'en', 'fr', 'de' ];
+const localesWithCookiePolicy = [ 'en', 'fr', 'de' ];
+
+const setLocalizedUrlHost = ( hostname, validLocales = [] ) => ( urlParts, localeSlug ) => {
+	const localesToSubdomains = {
+		'pt-br': 'br',
+		br: 'bre',
+		zh: 'zh-cn',
+		'zh-hk': 'zh-tw',
+		'zh-sg': 'zh-cn',
+		kr: 'ko',
+	};
+
+	if ( typeof validLocales === 'string' ) {
+		validLocales = config( validLocales );
 	}
 
-	return url;
+	if ( validLocales.includes( localeSlug ) ) {
+		urlParts.host = `${ localesToSubdomains[ localeSlug ] || localeSlug }.${ hostname }`;
+	}
+	return urlParts;
+};
+
+const prefixLocalizedUrlPath = ( validLocales = [] ) => ( urlParts, localeSlug ) => {
+	if ( typeof validLocales === 'string' ) {
+		validLocales = config( validLocales );
+	}
+	if ( validLocales.includes( localeSlug ) ) {
+		urlParts.pathname = localeSlug + urlParts.pathname;
+	}
+	return urlParts;
+};
+
+const urlLocalizationMapping = {
+	'wordpress.com': setLocalizedUrlHost( 'wordpress.com', 'magnificent_non_en_locales' ),
+	'wordpress.com/tos/': setLocalizedUrlHost( 'wordpress.com', 'magnificent_non_en_locales' ),
+	'jetpack.com': setLocalizedUrlHost( 'jetpack.com', 'jetpack_com_locales' ),
+	'en.support.wordpress.com': setLocalizedUrlHost(
+		'support.wordpress.com',
+		'support_site_locales'
+	),
+	'en.blog.wordpress.com': setLocalizedUrlHost( 'blog.wordpress.com', localesWithBlog ),
+	'en.forums.wordpress.com': setLocalizedUrlHost( 'forums.wordpress.com', 'forum_locales' ),
+	'automattic.com/privacy/': prefixLocalizedUrlPath( localesWithPrivacyPolicy ),
+	'automattic.com/cookies/': prefixLocalizedUrlPath( localesWithCookiePolicy ),
+};
+
+export function localizeUrl( fullUrl, locale ) {
+	const localeSlug = locale || getLocaleSlug();
+	const urlParts = url.parse( String( fullUrl ) );
```

Comment:
@ramonjd I opted for converting to a string here to allow for `toString()` implementations.